### PR TITLE
Feature/calendar/delegation

### LINKED ISSUE #7085: Calendar delegation
According to
[http://sabre.io/dav/caldav-proxy/](http://sabre.io/dav/caldav-proxy/)
.. sabredav supports the delegation of calendars from the client-side via caldav

however in Nextcloud/OwnCloud this seems not implemented.
Do you plan to integrate this in the near future?

Ps: I managed to test it with [Baikal](https://github.com/sabre-io/Baikal)

### Steps to reproduce
1. Create multiple Users on NC
2. Apple iCal/Calendar Settings -> Accounts -> Delegations (dont know the english word)
3. Searches live on server for matched Users

The query server-side appears with something like:

`[06/Nov/2017:16:56:10 +0100] "REPORT /basedir/remote.php/dav/principals/users/MYUSER/calendar-proxy-write/ HTTP/1.1" 404 1604 "-" "Mac+OS+X/10.13 (17A405) CalendarAgent/399"
`

**Operating system**:
Ubuntu Server 16 LTS
**Web server:**
Apache2 latest today
**Database:**
mysql
**PHP version:**
7.0
**Server version:** (see your admin page)
NC12.0.3
**Calendar version:** (see the apps page)
Calendar 1.5.6
**Updated from an older installed version or fresh install:**
fresh

```
http://example.com/index.php/settings/integrity/failed says:
No errors have been found.
```

### PR DESCRIPTION
**Pull request by rullzer**
fixes #7085

### CODE REVIEW

File: apps/dav/lib/CalDAV/Proxy/Proxy.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	/** @var string */
+	protected $ownerId;
+	/** @var string */
+	protected $proxyId;
+	/** @var int */
+	protected $permissions;
+
+	public function __construct() {
+		$this->addType('ownerId', 'string');
```
Comment (georgehrke) [inline apps/dav/lib/CalDAV/Proxy/Proxy.php]: default type is string, so adding a type for ownerId and proxyId is not necessary

File: apps/dav/lib/Migration/Version1011Date20190806104428.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+			'length' => 64,
+		]);
+		$table->addColumn('permissions', Type::INTEGER, [
+			'notnull' => false,
+			'length' => 4,
+			'unsigned' => true,
+		]);
+
+		$table->setPrimaryKey(['id']);
+		$table->addUniqueIndex(['owner_id', 'proxy_id', 'permissions'], 'dav_cal_proxy_uidx');
```
Comment (georgehrke) [inline apps/dav/lib/Migration/Version1011Date20190806104428.php]: I guess we should add indices for proxy_id and owner_id, because we use that for querying.

**Review by georgehrke — CHANGES_REQUESTED** [review]
First quick dirty review, will check in depth later

File: apps/dav/lib/CalDAV/Proxy/Proxy.php (hunk 1 above)
Comment (rullzer) [reply apps/dav/lib/CalDAV/Proxy/Proxy.php]: Yes I know. But sill being explicit is nice IMO

File: apps/dav/lib/Migration/Version1011Date20190806104428.php (hunk 2 above)
Comment (rullzer) [reply apps/dav/lib/Migration/Version1011Date20190806104428.php]: Fair enough

**Comment by tcitworld** [comment]
Could you explain what's the motive behind this?

**Comment by georgehrke** [comment]
> Could you explain what's the motive behind this?

This will allow you to delegate all your calendars to someone else. Most common use-case is someone sharing their calendars with their assistant/secretary to allow them to send out and respond to invitations on their behalf.

**Comment by tcitworld** [comment]
Thanks, wasn't sure if it was the issue that you linked on top.

File: apps/dav/lib/CalDAV/Calendar.php [hunk 3]
```diff
@@ -174,9 +181,37 @@ public function getACL() {
...
+		$acl = $this->addProxies($acl);
+
+		return $acl;
+	}
+
+	public function addProxies(array $acl): array {
+		list($prefix, $name) = \Sabre\Uri\split($this->getOwner());
+		$proxies = $this->proxyMapper->getProxiesOf($name);
+
+		foreach ($proxies as $proxy) {
```
Comment (georgehrke) [inline apps/dav/lib/CalDAV/Calendar.php]: Note for myself: Remove, just add principal /calendar-user-write

**Comment by georgehrke** [comment]
61A4B590-7041-4254-9A12-47DCD40AC5CC-2

🎉🎉🎉

**Comment by georgehrke** [comment]
tldr of what's left:
- [ ] add ProxyMapper to Resource and Room principal backends
- [ ] tests

**Comment by georgehrke** [comment]
Let's see if Drone is happy now.

**Comment by rullzer** [comment]
Lets merge this!

**Comment by alex-s-team** [comment]
pretty excited about this feature, has this been released already (nextcloud 18 is not showing any delegates in ical)? how can this be activated?

**Comment by tcitworld** [comment]
@alex-s-team It's available in Nextcloud 17. It will work with CalDAV clients that support the feature, such as Apple Calendar or BusyCal.

**Comment by alex-s-team** [comment]
thanks very much for your answer, we are using apple calendar and there are no delegates showing up. do i have to activate the feature somehow via config or is the principal url different?

**Comment by tcitworld** [comment]
As this seems to be a setup issue I would like to ask you to raise your question in the forums (so that we don't spam people in this issue): https://help.nextcloud.com

If you wish support with setup issues from Nextcloud GmbH they offer this as part of the Nextcloud subscription. Learn more about this at https://nextcloud.com/enterprise/

**Comment by rotdrop** [comment]
Just for the record: here we find that this does not work with the OSX calendar app: #17053. At least it is not possible to find anyone to delegate the calendar to ...

**Comment by z3ky** [comment]
Maybe, the caldav trashbin #26083 #1662 has broken delegation #42051
